orangecontrib/text/widgets/utils/readers.py

```diff
@@ -0,0 +1,215 @@
+            parser = PDFParser(f)
+        doc = PDFDocument()
+        parser.set_document(doc)
+        doc.set_parser(parser)
+        doc.initialize('')
+        rsrcmgr = PDFResourceManager()
+        laparams = LAParams()
+        laparams.char_margin = 0.1
+        laparams.word_margin = 1.0
+        device = PDFPageAggregator(rsrcmgr, laparams=laparams)
+        interpreter = PDFPageInterpreter(rsrcmgr, device)
+        extracted_text = []
+
+        for page in doc.get_pages():
+            interpreter.process_page(page)
+            layout = device.get_result()
+            for lt_obj in layout:
+                if isinstance(lt_obj, LTTextBox) or isinstance(lt_obj, LTTextLine):
+                    extracted_text.append(lt_obj.get_text())
+        self.content = ' '.join(extracted_text)
+        return self.make_text_data()
+
+
+class XmlReader(Reader):
+    ext = ".xml"
+
+    def read_file(self):
+        with open(self.path) as markup:
+            soup = BeautifulSoup(markup.read(), "lxml")
+        self.content = soup.get_text()
+        return self.make_text_data()
+
+
+DefaultFormats = ("docx", "odt", "txt", "pdf", "xml")
+
+
+class TextScan:
+    def __init__(self, startdir, formats=DefaultFormats, report_progress=None):
+        self.startdir = startdir
+        self.formats = formats
+        self.report_progress = report_progress
+        self.cancelled = False
+
+    def run(self):
+        text_data = []
+        filescanner = scan(self.startdir)
+        patterns = ["*.{}".format(fmt.lower()) for fmt in self.formats]
+
+        def fnmatch_any(fname, patterns):
+            return any(fnmatch.fnmatch(fname.lower(), pattern)
+                       for pattern in patterns)
+
+        batch = []
+
+        for path in filescanner:
+            if fnmatch_any(path, patterns):
+                reader = Reader.get_reader(path)
+                textdata = reader.read()
+                if textdata is not None:
+                    text_data.append(textdata)
+                    batch.append(text_data)
+
+            if self.cancelled:
+                return
+
+            if len(batch) == 10 and self.report_progress is not None:
+                self.report_progress(
+                    namespace(count=len(text_data),
+                              lastpath=text_data[-1].path,
+                              batch=batch))
+                batch = []
+
+        if batch and self.report_progress is not None:
+            self.report_progress(
+                    namespace(count=len(text_data),
+                              lastpath=text_data[-1].path,
+                              batch=batch))
+        return text_data
+
+    @classmethod
+    def get_categories(cls, text_data):
+        categories = []
+        for textdata in text_data:
+            categories.append(textdata.category)
+        return list(set(categories))
+
+def scan(topdir, include_patterns=("*",), exclude_patterns=(".*",)):
```

Comment:
How do I pass `include_patterns ` or `exclude_patterns ` from `TextScan` class; i.e. how can the user scan directories and load only `txt` for example?

I this is not possible, why the fuss about this?